### PR TITLE
451 system tests skeleton

### LINKED ISSUE #451: System tests skeleton and basic test
- Create the test binary with a simple single test.
- Add github workflow to run it with every PR.

### LINKED ISSUE #496: Save system tests DB to S3
Data written by the test should be accessible for debugging.

### PR DESCRIPTION
**Pull request by itaiad200**
closes #451 
closes #496 

Introducing Nessie - system testing binary that executes against a running lakeFS instance.
It currently contains just a first simple system-test that commits a file and reads it.
Also a workflow is added that will run this test for all PRs.

### CODE REVIEW

File: .github/workflows/nessie.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Nessie
+on:
+  push:
+    branches:
+      -master
+  pull_request:
```
Comment (itaiad200) [inline .github/workflows/nessie.yaml:6]: Should be removed?
More time per PR but adds value.

Comment (itaiad200) [reply .github/workflows/nessie.yaml:6]: @ozkatz

File: Makefile [hunk 2]
```diff
@@ -15,6 +16,7 @@ GO_TEST_MODULES=$(shell $(GOCMD) list ./... | grep -v 'lakefs/api/gen/')
 
 LAKEFS_BINARY_NAME=lakefs
 LAKECTL_BINARY_NAME=lakectl
+NESSIE_BINARY_NAME=nessie
```
Comment (ozkatz) [inline Makefile:19]: love it :)

File: nessie/run.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+func Run(ctx context.Context, config Config) error {
+	// initialize the env/repo
+	logger := logging.FromContext(ctx)
+	bucketPath := fmt.Sprintf("%s/%s", config.BucketPath, time.Now().UTC().Format("2006/01/02/150405"))
+	logger.WithField("bucketPath", bucketPath).Info("Starting nessie run")
+
+	cl := client.NewHTTPClientWithConfig(strfmt.Default, &client.TransportConfig{
+		Host:     config.BaseURL,
+		BasePath: client.DefaultBasePath,
+		Schemes:  []string{"http"},
```
Comment (ozkatz) [inline nessie/run.go:41]: I'd make this configurable to support running against a "real" system that might not allow http?

File: nessie/run.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	})
+
+	expBackoff := backoff.NewExponentialBackOff()
```
Comment (ozkatz) [inline nessie/run.go:44]: not sure this requires exponential backoff. Sometimes a for-loop is enough :)

File: nessie/run.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	logger.WithField("accessKeyID", res.Payload.AccessKeyID).
+		WithField("accessSecretKey", res.Payload.AccessSecretKey).
+		Info("Cluster setup successfully")
+
+	c, err := api.NewClient(fmt.Sprintf("http://%s/%s", config.BaseURL, client.DefaultBasePath), res.Payload.AccessKeyID, res.Payload.AccessSecretKey)
+	if err != nil {
+		return fmt.Errorf("failed to setup client: %w", err)
+	}
+
+	if err := c.CreateRepository(ctx, &models.RepositoryCreation{
```
Comment (ozkatz) [inline nessie/run.go:75]: Kind of a big function, dealing with 2 separate concerns - setting up an environment to work with, and starting from this line, the actual test itself.

I suggest modeling it using Go's test framework - you get reporting, coverage, etc out of the box.
A running system is private case for a fixture.

With `go test` you still get the benefits you wanted but it requires no special infra to setup, just a `go test -tags system`.
you'll also get the benefit of `go test -race` which should deterministically fail many cases where you'd otherwise get flakiness, so easier to reproduce as well.

File: nessie/run.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	logger.WithField("repo", repo).Info("Created first repository")
+
+	// TODO(itai): run the tests in parallel
+	if err := singleCommit(ctx, c); err != nil {
+		return err
+	}
+
+	logger.Info("All tests finished successfully")
+	return nil
```
Comment (ozkatz) [inline nessie/run.go:90]: less of an issue in CI, but locally, if this was a Go test case, you'd get the benefit of [t.Cleanup](https://golang.org/pkg/testing/#T.Cleanup)

File: Makefile (hunk 2 above)
Comment (itaiad200) [reply Makefile:19]: :)

File: go.mod [hunk 7]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/apache/thrift v0.13.0
 	github.com/aws/aws-sdk-go v1.34.0
 	github.com/benbjohnson/clock v1.0.3
+	github.com/cenkalti/backoff v2.2.1+incompatible
```
Comment (arielshaqed) [inline go.mod:10]: This is a scary dependency (because it says "`+incompatible`", and because we have `v3.2.2` below).  Can we use just a single version?

File: nessie/system_test.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	baseURL string
+
+	// rawBucketPath is the full path to the s3 bucket
+	rawBucketPath string
+
+	enabled bool
+	https   bool
+}
+
+var (
```
Comment (arielshaqed) [inline nessie/system_test.go:38]: I am uncomfortable with these global variables.  E.g. if we nested all tests inside a single function (and used `t.Run`) they could be local.  E.g. if they were attached to a single object we could construct it and pass _that_ around.
Rationale: the usual "global variables are only convenient in the short term" argument.  Consider how this scales to add more tests.  Even `client` might be pointed at different clients each time.

File: nessie/system_test.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+	config testsConfig
+	logger logging.Logger
+	client api.Client
+)
+
+const (
+	maxSetup     = 5 * time.Minute
```
Comment (arielshaqed) [inline nessie/system_test.go:45]: May as well make these flags too.

File: nessie/system_test.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	masterBranch = "master"
+)
+
+func init() {
+	flag.StringVar(&config.baseURL, "endpoint-url", "localhost:8000", "URL endpoint of the lakeFS instance")
+	flag.StringVar(&config.rawBucketPath, "bucket", "s3://nessie-system-testing", "Bucket's path")
+	flag.BoolVar(&config.enabled, "system-tests", false, "Enabling the system-tests")
+	flag.BoolVar(&config.https, "https", false, "Use HTTPS to connect to lakeFS")
```
Comment (arielshaqed) [inline nessie/system_test.go:53]: Maybe unify it with `endpoint-url` (which currently isn't really a URL)?

File: nessie/system_test.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+		os.Exit(0)
+	}
+
+	// initialize the env/repo
+	logger = logging.Default()
+	logger.WithField("bucketPath", config.rawBucketPath).Info("Starting nessie run")
+
+	schemas := []string{"http"}
+	if config.https {
+		schemas = append(schemas, "https")
```
Comment (arielshaqed) [inline nessie/system_test.go:70]: `https` should be before `http`.

File: nessie/system_test.go [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		logger.WithError(err).Info("Setup failed")
+
+		select {
+		case <-ctx.Done():
+			panic("setup failed after all retries")
+		case <-time.After(5 * time.Second):
+		}
+	}
```
Comment (arielshaqed) [inline nessie/system_test.go:92]: Now `ctx` is somewhat poisoned, it will have a low timeout.  So if used downstream it may blow up occasionally.  Could we either scope it to here or make it `nil` or give it a more suggestive name like `setupCtx`?

File: nessie/system_test.go [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+		schemas = append(schemas, "https")
+	}
+	cl := genclient.NewHTTPClientWithConfig(strfmt.Default, &genclient.TransportConfig{
+		Host:     config.baseURL,
+		BasePath: genclient.DefaultBasePath,
+		Schemes:  schemas,
+	})
+
+	ctx, _ := context.WithTimeout(context.Background(), maxSetup)
+	for {
```
Comment (arielshaqed) [inline nessie/system_test.go:79]: I believe this should be a separate step in an external tool.  Otherwise we need to copy it into every similar system.

File: nessie/system_test.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+			DisplayName: &adminUserName,
+		},
+		Context: ctx,
+	})
+	if err != nil {
+		panic(fmt.Errorf("lakefs setup request failed: %w", err))
+	}
+
+	logger.WithField("accessKeyID", res.Payload.AccessKeyID).
+		WithField("accessSecretKey", res.Payload.AccessSecretKey).
```
Comment (arielshaqed) [inline nessie/system_test.go:106]: Please don't log secrets.

File: nessie/system_test.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+		Info("Cluster setup successfully")
+
+	client, err = api.NewClient(fmt.Sprintf("http://%s/%s", config.baseURL, genclient.DefaultBasePath), res.Payload.AccessKeyID, res.Payload.AccessSecretKey)
```
Comment (arielshaqed) [inline nessie/system_test.go:109]: How does this respect `--https`?

File: .github/workflows/nessie.yaml [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+        run: docker-compose up -d
+        env:
+          TAG: dev
+          LAKEFS_BLOCKSTORE_TYPE: s3
+      - name: Run nessie
+        run: make nessie
+        env:
+          API_ENDPOINT: localhost:8000
+      - name: Export DB
+        if: ${{ always() }}
```
[the remark this answers is not in the document]
Comment (itaiad200) [reply .github/workflows/nessie.yaml:35]: Nope, steps don't run by default after one step has failed.

File: go.mod (hunk 7 above)
Comment (itaiad200) [reply go.mod:10]: Removed.

File: nessie/system_test.go (hunk 13 above)
Comment (itaiad200) [reply nessie/system_test.go:79]: What similar system? The idea is that you'll use this tool whenever you want to test the system.

File: nessie/system_test.go (hunk 14 above)
Comment (itaiad200) [reply nessie/system_test.go:106]: But you'll need them later on to debug the system if the tests failed..

Comment (itaiad200) [reply nessie/system_test.go:106]: Those secrets are for a system that is being destroyed after the tests, right?

File: docs/assets/js/swagger.yml [hunk 17]
```diff
@@ -90,7 +90,7 @@ definitions:
         type: string
         description: "Filesystem URI to store the underlying data in (i.e. 's3://my-bucket/some/path/')"
         example: "s3://example-bucket/"
-        pattern: '^(s3|mem|local|transient)://.*$'
+        pattern: '^(s3|gs|mem|local|transient)://.*$'
```
Comment (ozkatz) [inline docs/assets/js/swagger.yml:93]: I think this belongs to another PR?

File: nessie/system_test.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	flag.DurationVar(&config.maxSetup, "max-setup", 5*time.Minute, "Maximum time to wait for lakeFS setup")
+	flag.BoolVar(&config.enabled, "system-tests", false, "Enabling the system-tests")
+	flag.BoolVar(&config.https, "https", false, "Use HTTPS to connect to lakeFS")
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	logger = logging.Default()
+
+	if !config.enabled {
```
Comment (ozkatz) [inline nessie/system_test.go:63]: There's a standard way to do this with build tags

File: .env [hunk 19]
```diff
@@ -0,0 +1,2 @@
+TAG=latest
+BLOCKSTORE_TYPE=local
```
Comment (ozkatz) [inline .env:2]: I think these need to be prefixed with `LAKEFS_` to get picked up by viper.. Additionally I'd add
`LAKEFS_STATS_ENABLED=false` if this is to be used by local dev environments

Comment (itaiad200) [reply .env:2]: Done. Adding the LAKEFS_STATS_ENABLED flag to the workflow since the .env is used by default, so it will be disabled for all.

File: docs/assets/js/swagger.yml (hunk 17 above)
Comment (itaiad200) [reply docs/assets/js/swagger.yml:93]: Yeah, not sure why this appears here as it's already in master. Will use rebase instead of merge from now on.
